### ISSUE TITLE
Visualization of SyN's deformation field

### ISSUE THREAD

**Issue by tzayuan** [junyuchen245/TransMorph_Transformer_for_Medical_Image_Registration#14]
Hi, @junyuchen245 

I'm sorry to trouble you, I have some questions on how visual the deformation flow generated by SyN algotirhm.

My thought is to save the deformation flow file and visual this flow by using ITK-SNAP, but the result seems have something wrong.

This is my main code for saving deformation flow:
`f_img_set = sitk.ReadImage("./fixed.nii.gz")`
`f_img = ants.image_read("./fixed.nii.gz")`
`m_img = ants.image_read("./moving.nii.gz")`
`mytx = ants.registration(fixed=f_img, moving=m_img, type_of_transform='SyN')`
`flow = np.array(nib_load(mytx['fwdtransforms'][0]), dtype='float32', order='C')`
`flow_save = torch.from_numpy(flow).to(device).float()`
`save_image(flow_save.permute(3, 0, 1, 2, 4)[np.newaxis, ...], f_img_set, fileName + "_flow.nii.gz")`
where Function `save_image()` is defined as follows:
`
def save_image(img, ref_img, name):`
&emsp; &emsp; `img = sitk.GetImageFromArray(img[0, 0, ...].cpu().detach().numpy())`
&emsp; &emsp; `    img.SetOrigin(ref_img.GetOrigin())`
&emsp; &emsp; `    img.SetDirection(ref_img.GetDirection())`
&emsp; &emsp; `    img.SetSpacing(ref_img.GetSpacing())`
&emsp; &emsp; `    sitk.WriteImage(img, os.path.join('./result/', name))`

After this operation, I found that the saved deformation flow(flow.nii.gz) is a little bit strange, just like:
<img width="840" alt="Snipaste_2022-04-04_19-03-27" src="https://user-images.githubusercontent.com/25738822/161530899-edd3b383-7617-4fa3-a4e8-bf10d0d18a76.png">
So, nextly, I try to save this deformation flow, and try to warp the moving image by using this saved flow and STN function(I am confident that this function will work correctly for deformable flow generated by deep learning-based methods), but I found that the generated warped image is different  from warped image generated by ANTs, just like:
<img width="516" alt="Snipaste_2022-04-04_19-01-20" src="https://user-images.githubusercontent.com/25738822/161530603-1aa16f2f-8564-4fc1-ad1d-582b0ccc3eb7.png">
Left is warped image generated by saved deformation flow, right is warped image generated by ANTs function `ants.apply_transforms`

I have no idea what the problem is although I have tried many methods, so I come to you for help, thank you!

Yours, 
Tzayuan

**Comment by junyuchen245** [comment]
Hi @tzayuan ,

As this is unrelated to my scripts, I'll have to lower the priority of your question since I'm pressed for time for multiple tasks. I'll try to get back to you once I have more time to take a closer look at this.

Junyu

**Comment by junyuchen245** [comment]
Hi @tzayuan ,

Here is the code that I used for SyN visualization:
```python
x = data[0].squeeze(0).squeeze(0).detach().cpu().numpy()
y = data[1].squeeze(0).squeeze(0).detach().cpu().numpy()
x_seg = data[2].squeeze(0).squeeze(0).detach().cpu().numpy()
y_seg = data[3].squeeze(0).squeeze(0).detach().cpu().numpy()
x = ants.from_numpy(x)
y = ants.from_numpy(y)

x_ants = ants.from_numpy(x_seg.astype(np.float32))
y_ants = ants.from_numpy(y_seg.astype(np.float32))
start = time.time()
reg12 = ants.registration(y, x, 'SyNOnly', reg_iterations=(160, 80, 40), syn_metric='meansquares')
def_seg = ants.apply_transforms(fixed=y_ants,
                 moving=x_ants,
                 transformlist=reg12['fwdtransforms'],
                 interpolator='nearestNeighbor',)
end = time.time()
print(end - start)
def_out = ants.apply_transforms(fixed=y,
                 moving=x,
                 transformlist=reg12['fwdtransforms'],)

flow = np.array(nib_load(reg12['fwdtransforms'][0]), dtype='float32', order='C')
flow = flow[:,:,:,0,:].transpose(3, 0, 1, 2)

grid_img = mk_grid_img(8, 1)
grid_img = ants.from_numpy(grid_img.astype(np.float32))
def_grid = ants.apply_transforms(fixed=y,
                  moving=grid_img,
                  transformlist=reg12['fwdtransforms'], )
defout = def_grid.numpy()[:,:, slice_num]
plt.figure()
plt.imshow(defout, cmap='gray', vmin=0, vmax=1)
plt.axis('off')
plt.show()

flow_as_rgb(flow, slice_num)

def mk_grid_img(grid_step, line_thickness=1):
    grid_img = np.zeros((160, 192, 224))
    for j in range(0, grid_img.shape[1], grid_step):
        grid_img[:, j+line_thickness-1, :] = 1
    for i in range(0, grid_img.shape[0], grid_step):
        grid_img[i+line_thickness-1, :, :] = 1
    return grid_img

def flow_as_rgb(flow, slice_num):
    flow = flow[:, :, :, slice_num]
    flow_rgb = np.zeros((flow.shape[1], flow.shape[2], 3))
    for c in range(3):
        flow_rgb[..., c] = flow[c, :, :]
    lower = np.percentile(flow_rgb, 2)  # 算分位数
    upper = np.percentile(flow_rgb, 98)
    flow_rgb[flow_rgb < lower] = lower
    flow_rgb[flow_rgb > upper] = upper
    flow_rgb = (((flow_rgb - flow_rgb.min()) / (flow_rgb.max() - flow_rgb.min())))
    plt.figure()
    plt.imshow(flow_rgb, vmin=0, vmax=1)
    plt.axis('off')
    plt.show()
    print(lower)
    print(upper)
```

In theory, the spatial transformation function for deep-learning-based methods and 'ants.apply_transforms' are the same. However, in practice, there are a variety of possible explanations for the difference you are seeing. For instance, the coordinate system ('ij' or 'xy') and whether or not the displacement should be normalized. In general, I believe it is preferable to warp images using the method's dedicated spatial transformation.

Junyu

**Comment by tzayuan** [comment]
Hi, @junyuchen245 

I am sorry that it has taken so long to give you feedback. The reason is that there is a problem with our computing server.

Thanks for your suggestions and codes. This has been very helpful and enlightening to me.

As you say, the setting of ordinate system should be careful, and I solved the problem of flow display by setting ordinate system rightly.

What's more, thank you very much for the reference code for RGB flow visualization.

Best regards,
Ztayuan
